### ISSUE TITLE
Request: Ability to pass CompilationOptions to Script.Compile()

### ISSUE THREAD

**Issue by tenor** [dotnet/roslyn#13650]
Currently, there is no way to specify a compilation option when calling `Script.Compile()` or `Script.RunAsync()`. So there is no way to specify allowUnsafe, checked, concurrentbuild, optimizationLevel and other compilation options for Interactive/Script mode.

Please add an overload to `.Compile()`, `.RunAsync()` and `.EvaluateAsync()` methods that accepts a `CSharpCompilationOptions|VisualBasicCompilationOptions` object and have it bubble all to way to [CreateSubmission](https://github.com/dotnet/roslyn/blob/a76370241899a0dcdfb522730cdfdcd9fb662551/src/Scripting/CSharp/CSharpScriptCompiler.cs#L53).

Also consider including a `[CSharpCompilationOptions|VBCompilationOptions].ScriptDefault` static property that returns the default complation options for scripts.

**Comment by tmat** [comment]
It is intentional that CompilationOptions are not settable. Only a small subset of them is applicable to scripts. We might consider adding AllowUnsafe, CheckOverflow and OptimizationLevel to ScriptOptions and passing them to the CompilationOptions under the hood. Other than these 3 settings I don't quite see why should we expose other options.  Not sure why you'd want to change the defaults.

**Comment by tenor** [comment]
I'd like to set `concurrentbuild=false` and `allowUnsafe=false`. The former should result in faster compilation times in a multi-user environment (although I haven't ran benchmarks), and the latter prevents compilation of unsafe code (and is false by default in the regular compiler).

There are some other options which I don't need but seem useful to specify:

The `assemblyIdentiyComparer` option looks like something that ought to be configurable, and the following options may come in handy if there is a need to compile scripts to binaries and save them to disk.
```deterministic
reportSuppressedDiagnostics
scriptClassName
moduleName
assemblyIdentityComparer
platform
generalDiagnosticOption
warningLevel
```

**Comment by tmat** [comment]
We don't support saving scripts to disk thru scripting APIs. You can, of course, create script compilations yourself and then emit to PE files.

**Comment by tenor** [comment]
Yes, you're right. Disregard my last list. Is the `DesktopAssemblyIdentityComparer` necessary for script compilation?

**Comment by jaredpar** [comment]
Do we want to consider allowing the deterministic setting as well?  That could allow for future optimizations around caching.

**Comment by tmat** [comment]
What kind of caching? Scripts are compiled in-memory and not saved to disk.

**Comment by jaredpar** [comment]
I was envisioning something a bit diff.  Essentially not bothering to compile the script if the inputs had an equivalent checksum.  May be overkill though.  Probably spend more time computing the key than it would take to compile the script.

**Comment by tmat** [comment]
You can already do that. The API allows you to get a delegate to the compiled script. You can reuse that delegate if the inputs are the same.

**Comment by tmat** [comment]
The fact that in-memory metadata have a random guid in them is irrelevant for the ability to reuse the delegate.

**Comment by Hoschi1982** [comment]
Hey there,

do you plan to make it possible to set the OptimizationLevel for the ScriptOptions?
Actually I can only build the script with the default OptimizationLevel=Debug.

It would be great if you can give me a hint.

Thanks a lot
